tools/enrollment_preprocessor.py

```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env python3
+
+import csv
+import sys
+from argparse import ArgumentParser
+from collections import defaultdict
+from dataclasses import dataclass
+from datetime import datetime
+from io import BytesIO
+from itertools import chain
+from pathlib import Path
+from typing import Iterator, NamedTuple, TextIO
+
+from openpyxl import Workbook, load_workbook
+from openpyxl.cell import Cell
+
+
+@dataclass
+class User:
+    title: str
+    last_name: str
+    first_name: str
+    email: str
+
+    def full_name(self) -> str:
+        prefix = ""
+        if self.title:
+            prefix = f"{self.title} "
+        return f"{prefix}{self.first_name} {self.last_name}, {self.email}"
+
+
+class UserCells(NamedTuple):
+    title: Cell | None
+    last_name: Cell
+    first_name: Cell
+    email: Cell
+
+    def _clean(self) -> Iterator[str]:
+        for field in iter(self):
+            if not field or not field.value:
+                yield ""
+                continue
+            field.value = str(field.value).strip()
+            yield field.value
+
+    def clean_user(self):
+        return User(*self._clean())
+
+
+def user_from_row(row: tuple[Cell, ...]):
+    return [
+        UserCells(None, *row[:3]).clean_user(),
+        UserCells(*row[7:]).clean_user(),
+    ]
+
+
+def make_bold(text: str) -> str:
+    return f"\033[1m{text}\033[0m"
+
+
+def csv_users_by_email(user_data: TextIO) -> dict[str, User]:
+    users = {}
+    reader = csv.reader(user_data, delimiter=";", lineterminator="\n")
+    next(reader)  # skip header
+    for row in reader:
+        user = User(*row)
+        users[user.email] = user
+    return users
+
+
+def get_user_decisions(database_users: dict[str, User], workbook: Workbook) -> dict[str, User]:
+    conflicts_by_field: dict[str, list[User]] = defaultdict(list)
+    for sheet in workbook.worksheets:  # parse enrollment data and group conflicts
+        for imported in chain.from_iterable(map(user_from_row, sheet.iter_rows(min_row=2, min_col=2))):
+            if not imported.email:
+                continue
+            existing = database_users.setdefault(imported.email, imported)
+            if imported == existing:
+                continue
+            for field in ["title", "last_name", "first_name"]:
+                field_value = getattr(imported, field)
+                if field_value is not None and getattr(existing, field) != getattr(imported, field):
+                    conflicts_by_field[field].append(imported)
```

Comment:
The first two lines are superfluous and basically just repeat the individual tests from below, right? I would remove them then:
```suggestion
            for field in ["title", "last_name", "first_name"]:
                field_value = getattr(imported, field)
                if field_value is not None and getattr(existing, field) != getattr(imported, field):
                    conflicts_by_field[field].append(imported)
```